vpr/src/place/initial_placement.cpp

```diff
@@ -97,6 +97,33 @@ void print_sorted_blocks(const std::vector<ClusterBlockId>& sorted_blocks, const
 static void place_all_blocks(const std::vector<ClusterBlockId>& sorted_blocks,
                              enum e_pad_loc_type pad_loc_type);
 
+/**
+ * @brief If any blocks are unplaced after initial placement, this routine
+ * prints an error message showing the names, types, and IDs of the unplaced blocks
+ */
+static void print_unplaced_blocks();
+
+static void print_unplaced_blocks() {
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+    auto& place_ctx = g_vpr_ctx.placement();
+
+    int unplaced_blocks = 0;
+
+    for (auto blk_id : cluster_ctx.clb_nlist.blocks()) {
+        if (place_ctx.block_locs[blk_id].loc.x == INVALID_X) {
+            VTR_LOG("Block %s (# %d) of type %s could not be placed during initial placement\n", cluster_ctx.clb_nlist.block_name(blk_id).c_str(), blk_id, cluster_ctx.clb_nlist.block_type(blk_id)->name);
```

Comment:
Also print its floorplan constraint: say whether constrained or not, and what the constraint is.
Can also say if it is part of a placement macro, and if it is, give the other macro members and offsets.